src/cargo/core/resolver/version_prefs.rs

```diff
@@ -84,7 +84,38 @@ impl VersionPreferences {
                 return previous_cmp;
             }
 
-            if self.max_rust_version.is_some() {
+            if let Some(max_rust_version) = &self.max_rust_version {
+                match (a.rust_version(), b.rust_version()) {
+                    // Fallback
+                    (None, None) => {}
+                    (Some(a), Some(b)) if a == b => {}
```

Comment:
Can be combined with the previous as `(a, b) if a == b =>`

Reply:
The compiler doesn't recognize that `None == None` and complains about the `match` being non-exhaustive